File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.py
import re

def solution(new_id):
    # 1단계
    new_id = new_id.lower()
    
    # 2단계
    new_id = ''.join(re.sub("[^a-z0-9-_.]", "", new_id))

    # 3단계
    while ".." in new_id:
        new_id = new_id.replace("..", ".")
    
    # 4단계
    if len(new_id) > 1:
        if new_id[0] == ".":
            new_id = new_id[1:]
        if new_id[-1] == ".":
            new_id = new_id[:-1]
    if len(new_id) == 1 and new_id[0] == '.':
        new_id = ""
                     
    # 5단계
    if new_id == "":
        new_id += "a"
                    
    # 6단계
    if len(new_id) >= 16:
        new_id = new_id[:15]
        if new_id[-1] == ".":
            new_id = new_id[:-1]
    
    # 7단계
    if len(new_id) <= 2:
            while len(new_id) != 3:
                new_id += new_id[-1]
    
    return new_id

File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.java
class Solution {
    public String solution(String new_id) {
        // 1단계
        new_id = new_id.toLowerCase();
        
        // 2단계
        new_id = new_id.replaceAll("[^a-z0-9-_.]", "");
        
        // 3단계
        while (new_id.contains("..")) {
            new_id = new_id.replace("..", ".");
        }
        
        // 4단계
        if (new_id.length() > 0 && new_id.charAt(0) == '.') {
            new_id = new_id.substring(1);
        }
        if (new_id.length() > 0 && new_id.charAt(new_id.length() - 1) == '.') {
            new_id = new_id.substring(0, new_id.length() - 1);
        }
        
        // 5단계
        if (new_id.isEmpty()) {
            new_id = "a";
        }
        
        // 6단계
        if (new_id.length() >= 16) {
            new_id = new_id.substring(0, 15);
            if (new_id.charAt(new_id.length() - 1) == '.') {
                new_id = new_id.substring(0, 14);
            }
        }
        
        // 7단계
        if (new_id.length() <= 2) {
            while (new_id.length() != 3) {
                new_id += new_id.charAt(new_id.length() - 1);
            }
        }
        
        return new_id;
    }
}